File: Programmers/Level_3/네트워크.py
def solution(n, computers):
    answer = 0
    parent = [0] * (n+1) 
    visited = [False] * (n+1)

    for i in range(n+1):
        parent[i] = i 
    
    def find_parent(x):
        if parent[x] != x:
            parent[x] = find_parent(parent[x])
        return parent[x]
    
    def union_parent(a, b):
        a = find_parent(a)
        b = find_parent(b)
        if a < b:
            parent[b] = a 
        else:
            parent[a] = b 
    
    for i in range(len(computers)):
        for j in range(len(computers)):
            if computers[i][j]:
                union_parent(i+1, j+1)

    # 집합 개수 카운트 
    for i in range(1, n+1):
        if visited[i]: 
            continue 
        visited[i] = True 
        answer += 1
        for j in range(i+1, n+1):
            # i와 연관된 모든 노드들은 한 잡합 안에 있다고 체크 
            if find_parent(i) == find_parent(j): 
                visited[j] = True

    return answer

solution(4, [[1, 1, 0, 1], [1, 1, 0, 0], [0, 0, 1, 1], [1, 0, 1, 1]])

File: Programmers/Level_3/Swift_Code/Swift_Code/네트워크.swift
//
//  네트워크.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/05/25.
//

/*
 https://programmers.co.kr/learn/courses/30/lessons/43162
 
 인접리스트 생성 후 dfs로 이어진 노드를 모두 방문하면서 방문처리
 방문처리 안 된 노드가 있다면 해당 노드를 기점으로 다시 dfs 처리하면서 답 1 증가 
 */

import Foundation

func solution(_ n:Int, _ computers:[[Int]]) -> Int {
    var answer = 0
    var graph = [[Int]](repeating: [Int](), count: computers.count)
    var visited = [Bool](repeating: false, count: computers.count)
    
    func makeGraph() {
        for i in 0..<computers.count {
            for j in 0..<computers[0].count {
                if computers[i][j] == 1 && i != j {
                    graph[i].append(j)
                }
            }
        }
    }
    
    func dfs(_ idx: Int) {
        if visited[idx] == true {
            return
        }
        
        visited[idx] = true
        
        for e in graph[idx] {
            dfs(e)
        }
    }
    
    makeGraph()
    
    for i in 0..<computers.count {
        if visited[i] == false {
            answer += 1
            dfs(i)
        }
    }
    
    return answer
}

// print(solution(3, [[1, 1, 0], [1, 1, 1], [0, 1, 1]]))
